### PR TITLE
Restore "Fix defining multiple methods with the same block"

### PR DESCRIPTION
**Pull request by elia**
I ended up using `Proxy` and `Reflect` as they're supported everywhere except for explorer and seem to solve the issue pretty well.

Wasn't able to write a test that broke on this, I manually tested by adding paggio to the Gemfile and running:

```
bin/opal --gem paggio -rpaggio -e 'puts Paggio.html! { div.data(test: "test") {div.data(test: "test")} }'
```

---

This reverts commit 787c6065ff6045c57aeb3c98e887c920fb1270c7,
restoring commit 3eacd508eb55daa34198330bf6820579428d845d.

Ref https://github.com/opal/opal/pull/2377
Fixes https://github.com/opal/opal/pull/2379

### CODE REVIEW

File: opal/corelib/module.rb [hunk 1]
```diff
@@ -396,14 +396,39 @@ def define_method(name, method = undefined, &block)
...
 
     %x{
-      var id = '$' + name;
+      var blockOrWrapper;
 
-      block.$$jsid        = name;
-      block.$$s           = null;
-      block.$$def         = block;
-      block.$$define_meth = true;
+      if (block.$$jsid && Proxy && Reflect) {
```
Comment (hmdne) [inline opal/corelib/module.rb:401]: Proxy/Reflect if they don't exist they will throw an exception

**Comment by elia** [comment]
@hmdne you ever saw this error by any chance?

```
#
# Fatal error in , line 0
# Check failed: FLAG_fuzzing.
#
#
#
#FailureMessage Object: 0x7ffeb2b90d10
 1: 0x563689cee665  [node]
 2: 0x56368aa141a2 V8_Fatal(char const*, ...) [node]
 3: 0x56368a3614e9  [node]
 4: 0x56368a367010  [node]
 5: 0x56368a3671d9 v8::internal::Runtime_OptimizeFunctionOnNextCall(int, unsigned long*, v8::internal::Isolate*) [node]
 6: 0x56368a6f8ef4  [node]
```

**Comment by hmdne** [comment]
Never. Could it be due to usage of Proxy?

**Comment by elia** [comment]
Got to the bottom of it, functions wrapped in proxies can't be checked for opt status directly, opened an issue on V8 here: https://bugs.chromium.org/p/v8/issues/detail?id=12736

**Comment by elia** [comment]
Slight decrease in perf, but nothing extreme, we might be better off avoiding define method for meta programmed definitions:

```
Comparison of the Asciidoctor (a real-life Opal application) compile and run:
[152](https://github.com/opal/opal/runs/5689890553?check_suite_focus=true#step:11:152)
                  Compile time: 6.188 (±2.26%) -> 6.353 (±6.59%) (change: +2.66%)
[153](https://github.com/opal/opal/runs/5689890553?check_suite_focus=true#step:11:153)
                      Run time: 1.588 (±2.23%) -> 1.599 (±10.69%) (change: +0.65%)
[154](https://github.com/opal/opal/runs/5689890553?check_suite_focus=true#step:11:154)
                   Bundle size: 4745.37 kB -> 4746.22 kB (change: -1.00%)
[155](https://github.com/opal/opal/runs/5689890553?check_suite_focus=true#step:11:155)
          Minified bundle size: 1018.35 kB -> 1018.52 kB (change: -1.00%)
[156](https://github.com/opal/opal/runs/5689890553?check_suite_focus=true#step:11:156)
            Mangled & minified: 710.59 kB -> 710.72 kB (change: -1.00%)
```

**Pull request merged**